File: src/application/c-c++/set-admetadata-in-pipeline/classification/set_classification_in_pipeline.cpp
// **
// Demo senario: appsrc ! adroi_draw ! videoconvert ! ximagesink
// 
// This example only show how to feed frame data to appsrc and set classifications to adlink metadata version 2.
// So this example does not deal with any other detail concern about snchronize or other tasks.
// Only show how to set the adlink metdata through appsrc for user who is interested with them.
// **
#include "opencv2/opencv.hpp"
#include <iostream>
#include <stdio.h>
#include <thread>
#include <chrono>
#include <vector>
#include <gst/gst.h>

#include <gstadroi_frame.h>
#include <gstadroi_batch.h>

using namespace cv;
using namespace std;

static GMainLoop *loop;
GstElement *pipeline, *videotestsrc, *filtercaps, *appsink, *appsrc, *drawer, *videoconvert, *ximagesink;
vector<Mat> grabVec;
int cols, rows, depth, channels;

static void cb_need_data(GstElement *appsrc, guint unused_size, gpointer user_data)
{
    static GstClockTime timestamp = 0;
    GstBuffer *buffer;
    guint size;
    GstFlowReturn ret;
    GstMapInfo map;

    size = cols * rows * channels;

    /* create new buffer */
    buffer = gst_buffer_new_allocate (NULL, size, NULL);
    gst_buffer_map(buffer, &map, GST_MAP_WRITE);

    /* wait for image data vector */
    if( grabVec.size() > 0 )
    {
        memcpy((guchar *)map.data, grabVec[0].data, gst_buffer_get_size(buffer));
        grabVec.erase(grabVec.begin());
    }
    gst_buffer_unmap(buffer, &map);
    
    /* set buffer timestamp */
    GST_BUFFER_PTS (buffer) = timestamp;
    GST_BUFFER_DURATION (buffer) = gst_util_uint64_scale_int (1, GST_SECOND, 2);
    timestamp += GST_BUFFER_DURATION (buffer);
    
    /* set random labels to buffer */
    GstElement *element = (GstElement *) appsrc;
    auto *f_meta = gst_buffer_acquire_adroi_frame_meta(buffer, (GstPad *)g_list_nth_data(element->pads, 0));
    if (f_meta == nullptr) 
        std::cout << "Can not get adlink ROI frame metadata" << std::endl;

    auto *b_meta = gst_buffer_acquire_adroi_batch_meta(buffer);
    if (b_meta == nullptr)
        std::cout << "Can not acquire adlink ROI batch metadata" << std::endl;
    
    std::vector<std::string> labels = {"water bottle", "camera", "chair", "person", "slipper", "mouse", "Triceratops", "woodpecker"};
    srand( time(NULL) );
    int index = rand() % labels.size();

    auto qrs = f_meta->frame->query("//");
    if(qrs[0].rois.size() > 0)
    {
        qrs[0].rois[0]->add_classification("sample-engine", "", (float)index / labels.size(), labels[index], index);
    }
    
    /* push buffer to appsrc */
    g_signal_emit_by_name (appsrc, "push-buffer", buffer, &ret);
    gst_buffer_unref (buffer);

    if (ret != GST_FLOW_OK) 
        g_main_loop_quit (loop);
}

static void free_pipeline()
{
    gst_element_set_state (pipeline, GST_STATE_NULL);
    gst_object_unref (GST_OBJECT (pipeline));
    g_main_loop_unref (loop);
}

static void establish_thread_pipeline()
{
    /* init GStreamer */
    gst_init (NULL, NULL);
    loop = g_main_loop_new (NULL, FALSE);

    /* setup pipeline */
    pipeline = gst_pipeline_new ("pipeline");
    appsrc = gst_element_factory_make ("appsrc", "appsrc");
    drawer = gst_element_factory_make("adroi_draw", "drawer");
    videoconvert = gst_element_factory_make("videoconvert", "videoconvert");
    ximagesink = gst_element_factory_make("ximagesink", "ximagesink");

    /* setup appsrc*/
    g_object_set (G_OBJECT (appsrc), "caps",
                  gst_caps_new_simple ("video/x-raw", "format", G_TYPE_STRING, "BGR",
                                       "width", G_TYPE_INT, 640, "height", G_TYPE_INT, 480,
                                       "framerate", GST_TYPE_FRACTION, 30, 1,NULL), NULL);

    gst_bin_add_many (GST_BIN (pipeline), appsrc, drawer, videoconvert, ximagesink, NULL);
    gst_element_link_many (appsrc, drawer, videoconvert, ximagesink, NULL);

    g_object_set (G_OBJECT (appsrc), "stream-type", 0, "format", GST_FORMAT_TIME, NULL);
    g_signal_connect (appsrc, "need-data", G_CALLBACK (cb_need_data), NULL);
 
    /* play */
    gst_element_set_state (pipeline, GST_STATE_PLAYING);

    while(true)
    {
        this_thread::sleep_for(chrono::milliseconds(10));
    }
    free_pipeline();
}

int main(int, char**)
{
    Mat frame;
 
    VideoCapture cap("videotestsrc ! video/x-raw, width=640, height=480, framerate=30/1 ! videoconvert ! appsink", CAP_GSTREAMER);
    
    if (!cap.isOpened()) 
    {
        cerr << "ERROR! Unable to open camera\n";
        system("pause");
        return -1;
    }

    cout << "Start grabbing" << endl;

    /* get frame size */
    cap.read(frame);
    cols = frame.cols;
    rows = frame.rows;
    depth = frame.depth();
    channels = frame.channels();
    thread pipethread(establish_thread_pipeline);
    
    while (true)
    {
        if (frame.empty()) 
        {
            cerr << "ERROR! blank frame grabbed\n";
            break;
        }
        
        grabVec.push_back(frame.clone());
        
        this_thread::sleep_for(chrono::milliseconds(10));

        cap.read(frame);
    }

    cap.release();
    frame.release();
    return 0;
}

File: src/application/Python/set-admetadata-in-pipeline/classification/set_classification_in_pipeline.py
## **
## Demo senario: appsrc ! adroi_draw ! videoconvert ! ximagesink
## 
## This example only show how to feed frame data to appsrc and set classifications to adlink metadata.
## So this example does not deal with any other detail concern about snchronize or other tasks.
## Only show how to set the adlink metdata through appsrc for user who is interested with them.
## **
import sys
import time
import numpy
import random
import threading

# Required to import to set ADLINK inference metadata
import adroi

import cv2
import gi
gi.require_version('Gst', '1.0')

from gi.repository import Gst, GObject, GLib
from queue import Queue

grabBuffer = Queue()
num_frames = 0

def need_data(src, length) -> Gst.FlowReturn:
    # wait for image data vector, grabBuffer, is not full
    while True:
        time.sleep(0.001)
        if grabBuffer.qsize() > 0:
            break
    
    global num_frames
    if grabBuffer.qsize() > 0:
        # get image data from vector
        buf = Gst.Buffer.new_allocate(None, length, None)
        buf.fill(0, grabBuffer.get().tostring())
        
        # set buffer timestamp
        buf.duration = 1/ 30 * Gst.SECOND
        buf.pts = buf.dts = int(num_frames * buf.duration)
        buf.offset = num_frames * buf.duration
        num_frames += 1
        
        # init metadata
        f_meta = adroi.gst_buffer_acquire_adroi_frame_meta(hash(buf), hash(src.pads[0]))
        if f_meta is None:
            print("Can not get adlink ROI frame metadata")
        
        b_meta = adroi.gst_buffer_acquire_adroi_batch_meta(hash(buf))
        if b_meta is None:
            print("Can not get adlink ROI batch metadata")
            
        qrs = f_meta.frame.query('//')
        if qrs is None or len(qrs) == 0:
            print("query is empty from frame meta in classifier_sample.")
            return self.srcpad.push(buff)
      
        # create random classification
        cls = []
        labels = ['water bottle', 'camera', 'chair', 'person', 'slipper', 'mouse', 'Triceratops', 'woodpecker']
        class_id = random.randrange(len(labels))
        class_prob = random.uniform(0, 1)
        
        # add data to buffer
        if len(qrs[0].rois) > 0:
            qrs[0].rois[0].add_classification('sample-engine', '', class_prob, labels[class_id], class_id)

        # push buffer to appsrc
        retval = src.emit('push-buffer', buf)

        if retval != Gst.FlowReturn.OK:
            print("retval = ", retval)

        time.sleep(0.01)

def enough_data(src, size, length) -> Gst.FlowReturn:
    return Gst.FlowReturn.OK

def on_message(bus: Gst.Bus, message: Gst.Message, loop: GObject.MainLoop):
    mtype = message.type
    
    if mtype == Gst.MessageType.EOS:
        print("End of stream")
        loop.quit()
    elif mtype == Gst.MessageType.ERROR:
        err, debug = message.parse_error()
        print("Gst.MessageType.ERROR catched in on_message:")
        print(err, debug)
        loop.quit()
    elif mtype == Gst.MessageType.ANY:
        err, debug = message.parse_warning()
        print(err, debug)

    return True

def establish_pipeline(pipeline, pipeline_elements):
    ## Add elements in pipeline.
    for element in pipeline_elements:
        pipeline.add(element)
    ## Link element one by one.
    for i in range(len(pipeline_elements) - 1):
        pipeline_elements[i].link(pipeline_elements[i + 1])

def establish_thread_pipeline():
    print('Start establish thread pipeline.')
    
    # Initialize GStreamer
    Gst.init(sys.argv)

    # Create the elements
    ## element: appsrc
    src = Gst.ElementFactory.make("appsrc", "src")
    caps = Gst.caps_from_string("video/x-raw,format=BGR,width=640,height=480,framerate=30/1")
    src.set_property('caps', caps)
    src.set_property('blocksize', 640*480*3)
    src.set_property('emit-signals', True)
    src.connect('need-data', need_data)
    src.connect('enough-data', enough_data)
    
    ## element: addraw_roi
    drawer = Gst.ElementFactory.make("adroi_draw", "drawer")

    ## element: videoconvert
    videoconvert = Gst.ElementFactory.make("videoconvert", "videoconvert")

    ## element: ximagesink
    sink = Gst.ElementFactory.make("ximagesink", "sink")
    
    # Create the empty pipeline
    pipeline = Gst.Pipeline().new("test-pipeline")
    
    # Build the pipeline
    pipeline_elements = [src, drawer, videoconvert, sink]
    establish_pipeline(pipeline, pipeline_elements)

    # Start pipeline
    pipeline.set_state(Gst.State.PLAYING)
    loop = GLib.MainLoop()

    # Wait until error or EOS
    bus = pipeline.get_bus()
    bus.add_signal_watch()
    bus.connect("message", on_message, loop)

    try:
        print("Start to run the pipeline in thread.")
        loop.run()
    except Exception:
        traceback.print_exc()
        loop.quit()

    # Stop Pipeline
    pipeline.set_state(Gst.State.NULL)
    del pipeline
    print('pipeline stopped.')

if __name__ == '__main__':

    pipthread = threading.Thread(target=establish_thread_pipeline)
    pipthread.daemon = True

    if sys.platform == 'win32' or sys.platform == 'linux':
        cap = cv2.VideoCapture("videotestsrc ! video/x-raw, width=640, height=480, framerate=30/1 ! videoconvert ! appsink", cv2.CAP_GSTREAMER)
    else:
        print("[VideoCapture] Do not support this system platform.")
        sys.exit()

    if cap.isOpened() == False:
        print("ERROR! Video frame capture failed")
        sys.exit()
      
    pipthread.start()
    while(True):
        ret, frame = cap.read()
        
        # push the image data to vector for the pipeline created in thread as a data provider
        grabBuffer.put(frame)
        
        time.sleep(0.05)
        
        if not pipthread.is_alive():
            break;

    cap.release()
    cv2.destroyAllWindows()

    pipthread.join()
    print('main thread end.')

